text/0517-io-os-reform.md

```diff
@@ -483,7 +483,72 @@ throughout IO, we can go on to explore the modules in detail.
 ### `std::env`
 [std::env]: #stdenv
 
-> To be added in a follow-up PR.
+Most of what's available in `std::os` today will move to `std::env`,
+and the signatures will be updated to follow this RFC's
+[Design principles] as follows.
+
+**Arguments**:
+
+* `args`: change to yield an iterator rather than vector if possible; in any case, it should produce an `OsStrBuf`.
```

Comment:
> Are you trying to say that the current implementation can't be improved this way? 

Ah yes, I'll try to be more specific. Today we do not store the value of `argv` passed to `main`, we make a local copy and stick it in a `static`. This is then deallocated when `std::rt::cleanup` is called. This means that the current implementation most definitely does not have a static lifetime on unix (minus ios/osx).

I believe the only lifetime which can be returned for a `CowOsString` is `'static` as we otherwise do not have a lifetime to tie to. This means that the current implementation cannot ever yield a borrowed instance (none have the static lifetime). An alternative implementation would be to perhaps store the raw value of `argv` in a global location to be referenced later. 

The primary reason I am worried about doing this is that I do not know the answer to: "Does this pointer actually have a static lifetime?" Arbitrary code can be running while `main` is running so if the arguments are stored anywhere locally and deallocated before exit this could lead to memory unsafety (via concurrent code in other threads).

Some other minor concerns (but not showstoppers) are:

* Using a raw `argv` pointers means that it's basically impossible to support the [`args::init`](http://doc.rust-lang.org/std/rt/args/fn.init.html) API. This is not necessarily a large worry, however.
* I have been worried about passing pointers into this memory as we don't necessarily have any guarantees that no one else is modifying it. You bring up a good point, however, that this also applicable to the implementation which copies out of the pointers. The upside of not passing the raw pointer out is that there is a deterministic scope in which you know that the arguments are accessed from the system, but I do not know if this would actually be relevant at all in practice. I doubt that anyone modifies arguments anyway.

Does that help clear up my concerns?